### PR TITLE
[pipeline]  All bert models

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
#4190 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
This pr finishes the pipeline builds for bert models 

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/test_shardformer/test_model/test_pure_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        #dataloader=prepare_dataloader(dataset=dataset['train'],batch_size=4)
+        schedule = OneForwardOneBackwardSchedule(num_microbatches, stage_manager)
+        shard_config = ShardConfig(enable_fused_normalization=enable_fused_normalization,
+                                   enable_tensor_parallelism=enable_tensor_parallelism,
+                                   pipeline_stage_manager=stage_manager)
+        pipelined_model = PipelinedModel(org_model, shard_config, stage_manager)
+        pp_optimizer = PipelineOptimizer(optimizer, pipelined_model)
+        data_it = iter(data_iter())
+        results = execute_pipeline(data_it, pipelined_model, loss, pp_optimizer, schedule=schedule)
+        assert results['loss'] is not None
```
Comment (Fridge003) [inline tests/test_shardformer/test_model/test_pure_pipeline.py:144]: Here the loss of org_model and pipelined_model should be asserted as equal

**Review by Fridge003 — CHANGES_REQUESTED** [review]

File: colossalai/shardformer/policies/bert.py [hunk 2]
```diff
@@ -297,9 +348,47 @@ def __init__(self) -> None:
...
+        bert_model = self.model.bert
+        if id(bert_model.embeddings.word_embeddings.weight) == id(self.model.cls.predictions.decoder.weight):
+            #tie weights
+            return [{
+                0: bert_model.embeddings.word_embeddings.weight,
+                self.pipeline_stage_manager.num_stages - 1: self.model.cls.predictions.decoder.weight
+            }]
+        return []
 
     def postprocess(self):
```
Comment (Fridge003) [inline colossalai/shardformer/policies/bert.py:393]: Should check whether pipeline parallel is used.

File: colossalai/shardformer/policies/bert.py [hunk 3]
```diff
@@ -237,7 +265,13 @@ def get_held_layers(self) -> List[Module]:
...
+        model = self.model
+        if id(model.bert.embeddings.word_embeddings.weight) == id(model.cls.predictions.decoder.weight):
+            #tie weights
+            return [{
+                0: model.bert.embeddings.word_embeddings.weight,
+                self.pipeline_stage_manager.num_stages - 1: model.cls.predictions.decoder.weight
+            }]
         return []
 
     def postprocess(self):
```
Comment (Fridge003) [inline colossalai/shardformer/policies/bert.py:277]: Should check whether pipeline parallel is used.

**Review by Fridge003 — CHANGES_REQUESTED** [review]

File: colossalai/shardformer/policies/bert.py [hunk 4]
```diff
@@ -278,7 +323,13 @@ def get_held_layers(self) -> List[Module]:
...
+        bert_model = self.model.bert
+        if id(bert_model.embeddings.word_embeddings.weight) == id(self.model.cls.predictions.decoder.weight):
+            #tie weights
+            return [{
+                0: bert_model.embeddings.word_embeddings.weight,
+                self.pipeline_stage_manager.num_stages - 1: self.model.cls.predictions.decoder.weight
+            }]
         return []
 
     def postprocess(self):
```
[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/shardformer/policies/bert.py:335]: When PP is used,  nothing should be done in postprocess

File: colossalai/pipeline/p2p.py [hunk 5]
```diff
@@ -64,7 +64,10 @@ def _broadcast_object_list(object_list: List[Any],
     my_rank = dist.get_rank()
     # Serialize object_list elements to tensors on src rank.
     if my_rank == src:
-        tensor_list, size_list = zip(*[c10d._object_to_tensor(obj) for obj in object_list])
+        if torch.__version__ >= "1.13.0":
+            tensor_list, size_list = zip(*[c10d._object_to_tensor(obj, device=device) for obj in object_list])
```
[the remark this answers is not in the document]
Comment (CjhHa1) [reply colossalai/pipeline/p2p.py:68]: in higher version, torch add a argument device , which need to be specified, so here we need a judgement, and this is updated in another pr

File: tests/test_pipeline/test_policy/test_bert_for_pretraining_model.py [hunk 6]
```diff
@@ -35,15 +36,20 @@ def check_bert_for_pretraining_forward():
...
-                                              attention_mask=attention_mask,
-                                              stage_manager=stage_manager)
+        output = bert_for_pretraining_forward(
+            self=model,
+            input_ids=x,
+            attention_mask=attention_mask,
+            stage_manager=stage_manager,
+            stage_index=stage_index,
+        )
         print(output['hidden_states'].shape)
```
[the remark this answers is not in the document]
Comment (CjhHa1) [reply tests/test_pipeline/test_policy/test_bert_for_pretraining_model.py:53]: done

File: tests/test_shardformer/test_model/test_pure_pipeline.py (hunk 1 above)
Comment (CjhHa1) [reply tests/test_shardformer/test_model/test_pure_pipeline.py:144]: This is done in another PR

File: tests/test_pipeline/test_policy/test_bert_model.py [hunk 7]
```diff
@@ -112,4 +118,3 @@ def test_bert_model_policy():
     """test the bert model forward and bert model policy"""
     #test_bert_model_forward()
     test_bert_model_policy()
```
[the remark this answers is not in the document]
Comment (CjhHa1) [reply tests/test_pipeline/test_policy/test_bert_model.py:121]: yes , it is true ,and I will finish it after Bert pipeline is tested

File: colossalai/shardformer/policies/bert.py (hunk 2 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/bert.py:376]: done

File: colossalai/shardformer/policies/bert.py (hunk 3 above)
Comment (CjhHa1) [reply colossalai/shardformer/policies/bert.py:276]: done

**Review by Fridge003 — APPROVED** [review]

**Comment by CjhHa1** [comment]
截屏2023-07-17 16 09 59
